[core] Prepend "use-client" directive + add docs and examples for using MUI libraries with Next.js App Router (#37656)
mui/material-ui: docs/data/base/guides/next-js-app-router/next-js-app-router.md

# Next.js App Router

<p class="description">Learn how to use Base UI with the Next.js App Router.</p>

:::info
Starting fresh on a new App Router-based project?

Jump right into the code with this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-ts).
:::

## Next.js and React Server Components

The Next.js App Router implements React Server Components, a [new feature](https://github.com/reactjs/rfcs/blob/main/text/0227-server-module-conventions.md#changes-since-v1) introduced in React 18.

> To support the App Router, currently all components and hooks from Base UI and other MUI libraries are designated as Client Components and exported with the `"use client"` directive.

Reviewer: I wonder if it's correct to say that **hooks** are client components

:::warning
React Server Components does not replace, and is separate from server-side rendering (SSR). Client Components are still server-rendered to HTML.

For more details, see this [explanation](https://github.com/reactwg/server-components/discussions/4) by the React team.
:::

## Setting up Base UI with the App Router

Base UI is styling-agnostic; setting up a Next.js App Router project largely depends on the styling solution:

### Emotion

> If you're using Emotion, or something Emotion-based like MUI System, you can follow the [same steps](/material-ui/guides/next-js-app-router/#using-material-ui-with-a-custom-theme) outlined for Material UI.

Reviewer 2: Maybe we should provide those instructions here in this doc too? Directing readers to the Material UI docs could still lead to confusion. For example, when I visit the page linked, it tells me to import a Material UI component (`ThemeProvider`), but I don't know if that component is also provided by Base UI, or if I have to install Material UI to keep going.

Author: OK – originally I just linked it because I didn't want CSS-in-JS to take up so much space 😓 

I'll simplify the Emotion setup by a little bit and move Tailwind up to be the first section

### Tailwind CSS

> [Install and initialize](https://tailwindcss.com/docs/guides/nextjs) Tailwind CSS, be sure to add the `app` directory and other directories to `tailwind.config.js`

Reviewer 2: To make the link more informative—readers should know exactly what they'll find if they click the link.

```suggestion
Follow the [Tailwind CSS doc on working with Next.js](https://tailwindcss.com/docs/guides/nextjs), and be sure to add the `app` directory and other directories to `tailwind.config.js`, as shown below:
```

```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    './src/app/**/*.{js,ts,jsx,tsx,mdx}',
    './src/components/**/*.{js,ts,jsx,tsx,mdx}'

    // or if not using the `src` directory:
    './app/**/*.{js,ts,jsx,tsx,mdx}',
  ],
  theme: {
    extend: {},
  },
  plugins: [],
};
```

Refer to this [example repo](https://github.com/mui/material-ui/blob/master/examples/base-next-app-router-tailwind-ts) for a full working demo of a Next.js 13 app using Base UI and Tailwind CSS.

> ### Other CSS-in-JS

Reviewer 2: ```suggestion
### Other CSS-in-JS libraries
```

> Base UI works with other CSS-in-JS libraries such as `styled-components`, see the [Next.js docs](https://nextjs.org/docs/app/building-your-application/styling/css-in-js) for more details.

Reviewer 2: ```suggestion
To use Next.js with Base UI and styled-components or other CSS-in-JS solutions, follow the [Next.js doc on CSS-in-JS](https://nextjs.org/docs/app/building-your-application/styling/css-in-js).
```

## Customization

### Using callbacks for slot props

> A common customization method in Base UI is to pass a callback to slots in `slotProps` in order to apply dynamic props to slots. For example, you might want to change the background color by applying a different class when a Button is disabled:

Reviewer: Too many "slots" ;)

```suggestion
A common customization method in Base UI is to pass a callback to slots in `slotProps` in order to apply dynamic props. For example, you might want to change the background color by applying a different class when a Button is disabled:
```

```tsx
// page.tsx

export default function Page() {
  return (
    <>
      {/* Next.js won't render this button without 'use-client'*/}
      <Button
        slotProps={{
          root: (ownerState: ButtonOwnerState) => ({
            className: ownerState.disabled ? 'bg-gray-400' : 'bg-blue-400',
          }),
        }}
      >
        Submit
      </Button>

      {/* Next.js can render this */}
      <Button
        slotProps={{
          root: {
            className: 'bg-gray-400',
          },
        }}
      >
        Return
      </Button>
    </>
  );
}
```

> Unfortunately, **this will not work in a server component** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).

Reviewer 2: Next.js and React seem to be going with title case for Server Component and Client Component, so we might as well follow the convention.

```suggestion
Unfortunately, **this does not work in a Server Component** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
```

> Unfortunately, **this will not work** since function props are [non-serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).

Reviewer: Let's underline that it won't work **in a server component** and it's still valid code with `"use client"`.
